### PR TITLE
⚠️  in-place propagation support for KCP

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

This PR implements the in-place propagation behavior for KCP.
The PR contains the following features:
- Changes to in-place propagation fields in KCP will not trigger a machine rollout
- Changes to in-place propagation fields in KCP are synced to the target Machines, InfraMachines and KubeadmConfigs
   - Example: A user can change`kcp.spec.machineTemplate.metadata.labels` without triggering a rollout and the changes will be passed down to the target Machines.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

Part of https://github.com/kubernetes-sigs/cluster-api/issues/7731

### CODE REVIEW

File: controlplane/kubeadm/internal/filters.go [hunk 1]
```diff
@@ -34,9 +34,9 @@ import (
 // Kubernetes version, infrastructure template, and KubeadmConfig field need to be equivalent.
 func MatchesMachineSpec(infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane) func(machine *clusterv1.Machine) bool {
 	return collections.And(
-		func(machine *clusterv1.Machine) bool {
-			return matchMachineTemplateMetadata(kcp, machine)
-		},
+		//func(machine *clusterv1.Machine) bool {
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/filters.go:37]: I think we still have to trigger a rollout if the cluster config annotation changed.

We use this annotation to store with which configuration we ran kubeadm init/join on the machine.

I think we can't simply skip rollout by in-place mutating this annotation

```go
	// Machine's bootstrap config may be missing ClusterConfiguration if it is not the first machine in the control plane.
	// We store ClusterConfiguration as annotation here to detect any changes in KCP ClusterConfiguration and rollout the machine if any.
```

File: controlplane/kubeadm/internal/controllers/helpers.go [hunk 2]
```diff
@@ -289,48 +292,140 @@ func (r *KubeadmControlPlaneReconciler) generateKubeadmConfig(ctx context.Contex
...
-		return errors.Wrap(err, "failed to marshal cluster configuration")
+func (r *KubeadmControlPlaneReconciler) computeDesiredMachine(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, cluster *clusterv1.Cluster, infraRef, bootstrapRef *corev1.ObjectReference, failureDomain *string, existingMachine *clusterv1.Machine) (*clusterv1.Machine, error) {
+	desiredMachine := &clusterv1.Machine{}
+
+	var machineName string
+	var machineUID types.UID
+	var machineFailureDomain *string
+	var version *string
+	annotations := map[string]string{}
+	if existingMachine == nil {
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/helpers.go:362]: Similar to the other PR. I think it would be easier to just construct the whole Machine with everything that is equal and then modify it later with if/else depending on the machine already exists

(also the diff becomes a lot smaller)

File: controlplane/kubeadm/internal/controllers/controller.go [hunk 3]
```diff
@@ -535,6 +571,72 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o client.Ob
 	return nil
 }
 
+func (r *KubeadmControlPlaneReconciler) adjustManagedFields(ctx context.Context, m *clusterv1.Machine) error {
```
Comment (ykakarap) [inline controlplane/kubeadm/internal/controllers/controller.go:574]: This will be dropped in favor of the ssa util package that will be introduced in https://github.com/kubernetes-sigs/cluster-api/pull/7921.

File: controlplane/kubeadm/internal/filters.go (hunk 1 above)
Comment (ykakarap) [reply controlplane/kubeadm/internal/filters.go:37]: That is covered under `MatchesKubeadmBootstrapConfig`. We dont need to do anything else here.

Comment (sbueringer) [reply controlplane/kubeadm/internal/filters.go:37]: Ah you're right!

**Comment by sbueringer** [comment]
General approach seems fine to me

**Comment by fabriziopandini** [comment]
@ykakarap I made a first pass and I have really appreciated the consistency of this work with other PRs of the same effort, also, the implementation is really straightforward, so I'm looking forward to see the unit test and making a final pass soon

File: controllers/external/util.go [hunk 4]
```diff
@@ -114,6 +117,84 @@ func CreateFromTemplate(ctx context.Context, in *CreateFromTemplateInput) (*core
...
+	// Labels is an optional map of labels to be added to the object.
+	// +optional
+	Labels map[string]string
+
+	// Annotations is an optional map of annotations to be added to the object.
+	// +optional
+	Annotations map[string]string
+
+	// Manager is the name of the manager to when using SSA patch.
+	Manager string
```
Comment (sbueringer) [inline controllers/external/util.go:156]:
```suggestion
	// FieldManager is the name of the field manager when using SSA patch.
	FieldManager string
```
nit

File: controlplane/kubeadm/internal/controllers/controller.go [hunk 5]
```diff
@@ -445,6 +436,57 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
...
+			continue
+		}
+		if err := ssa.CleanUpManagedFieldsForSSAAdoption(ctx, machine, kcpManagerName, r.Client); err != nil {
+			return errors.Wrapf(err, "failed to clean up managedFields of Machine %s", klog.KObj(machine))
+		}
+
+		fixedMachine := &clusterv1.Machine{}
+		if err := r.Client.Get(ctx, client.ObjectKeyFromObject(machine), fixedMachine); err != nil {
+			return err
+		}
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/controller.go:455]: unused

File: controlplane/kubeadm/internal/controllers/controller_test.go [hunk 6]
```diff
@@ -824,10 +1091,16 @@ func TestKubeadmControlPlaneReconciler_ensureOwnerReferences(t *testing.T) {
 
 	t.Run("replace non-KCP controller with KCP controller reference", func(t *testing.T) {
 		objs := []client.Object{fakeGenericMachineTemplateCRD, cluster.DeepCopy(), kcp.DeepCopy(), tmpl.DeepCopy()}
-		for _, purpose := range []secret.Purpose{secret.ClusterCA, secret.FrontProxyCA, secret.ServiceAccount, secret.EtcdCA} {
+		certificates := secret.Certificates{
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/controller_test.go:1094]: Let's definitely preserve those changes for envtest on a branch/PR somewhere, so we can use them later for the envtest migration

**Review by sbueringer — COMMENTED** [review]
Was just taking a quick look and left some comments that might stil be useful, even though the PR is WIP ;)

File: controlplane/kubeadm/internal/controllers/helpers.go [hunk 7]
```diff
@@ -288,60 +290,168 @@ func (r *KubeadmControlPlaneReconciler) generateKubeadmConfig(ctx context.Contex
...
+	machine, err := r.computeDesiredMachine(kcp, cluster, infraRef, bootstrapRef, failureDomain, nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to create Machine: failed to compute desired machine")
+	}
+	patchOptions := []client.PatchOption{
+		client.ForceOwnership,
+		client.FieldOwner(kcpManagerName),
+	}
+	if err := r.Client.Patch(ctx, machine, client.Apply, patchOptions...); err != nil {
+		return errors.Wrap(err, "failed to create Machine: Apply failed")
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/helpers.go:329]:
```suggestion
		return errors.Wrap(err, "failed to create Machine: apply failed")
```
nit: same below

File: controlplane/kubeadm/internal/controllers/helpers.go [hunk 8]
```diff
@@ -288,60 +290,168 @@ func (r *KubeadmControlPlaneReconciler) generateKubeadmConfig(ctx context.Contex
...
 		ObjectMeta: metav1.ObjectMeta{
-			Name:        names.SimpleNameGenerator.GenerateName(kcp.Name + "-"),
-			Namespace:   kcp.Namespace,
-			Labels:      internal.ControlPlaneMachineLabelsForCluster(kcp, cluster.Name),
-			Annotations: map[string]string{},
-			// Note: by setting the ownerRef on creation we signal to the Machine controller that this is not a stand-alone Machine.
+			UID:       machineUID,
+			Name:      machineName,
+			Namespace: kcp.Namespace,
 			OwnerReferences: []metav1.OwnerReference{
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/helpers.go:419]:
```suggestion
			// Note: by setting the ownerRef on creation we signal to the Machine controller that this is not a stand-alone Machine.
			OwnerReferences: []metav1.OwnerReference{
```
nit: let's keep the comment

File: controlplane/kubeadm/internal/controllers/helpers.go [hunk 9]
```diff
@@ -288,60 +290,168 @@ func (r *KubeadmControlPlaneReconciler) generateKubeadmConfig(ctx context.Contex
...
 	// Add the annotations from the MachineTemplate.
 	// Note: we intentionally don't use the map directly to ensure we don't modify the map in KCP.
 	for k, v := range kcp.Spec.MachineTemplate.ObjectMeta.Annotations {
-		machine.Annotations[k] = v
-	}
-	machine.Annotations[controlplanev1.KubeadmClusterConfigurationAnnotation] = string(clusterConfig)
-
-	if err := r.Client.Create(ctx, machine); err != nil {
-		return errors.Wrap(err, "failed to create machine")
+		annotations[k] = v
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/helpers.go:447]: Can we write those annotations here directly on the Machine and add the ones from `annotations` after that?

This ensures that "our" annotations cannot be overwritten by the ones from the MachineTemplate

File: controlplane/kubeadm/internal/controllers/helpers.go [hunk 10]
```diff
@@ -288,60 +290,168 @@ func (r *KubeadmControlPlaneReconciler) generateKubeadmConfig(ctx context.Contex
...
+		// For existing machine only set the ClusterConfiguration annotation if the machine already has it.
+		// We should not add the annotation if it was missing in the first place because we do not have enough
+		// information.
+		if clusterConfig, ok := existingMachine.Annotations[controlplanev1.KubeadmClusterConfigurationAnnotation]; ok {
+			annotations[controlplanev1.KubeadmClusterConfigurationAnnotation] = clusterConfig
+		}
+
+		// If the machine already has remediation data then preserve it.
+		// NOTE: This is required in order to track remediation retries.
+		if remediationData, ok := existingMachine.Annotations[controlplanev1.RemediationInProgressAnnotation]; ok {
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/helpers.go:404]:
```suggestion
		if remediationData, ok := existingMachine.Annotations[controlplanev1.RemediationForAnnotation]; ok {
```

File: controlplane/kubeadm/internal/controllers/helpers_test.go [hunk 11]
```diff
@@ -539,53 +555,132 @@ func TestKubeadmControlPlaneReconciler_generateMachine(t *testing.T) {
...
+
+		expectedMachineSpec := clusterv1.MachineSpec{
+			ClusterName: cluster.Name,
+			Version:     pointer.String(kcp.Spec.Version),
+			Bootstrap: clusterv1.Bootstrap{
+				ConfigRef: bootstrapRef.DeepCopy(),
+			},
+			InfrastructureRef:       *infraRef.DeepCopy(),
+			FailureDomain:           failureDomain,
+			NodeDrainTimeout:        kcp.Spec.MachineTemplate.NodeDrainTimeout.DeepCopy(),
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/helpers_test.go:578]: nit: Let's drop the deepcopy calls if we don't need them (same the next two lines and in other cases)

File: controlplane/kubeadm/internal/controllers/helpers_test.go [hunk 12]
```diff
@@ -539,53 +555,132 @@ func TestKubeadmControlPlaneReconciler_generateMachine(t *testing.T) {
...
+		g.Expect(createdMachine.Name).To(HavePrefix(kcp.Name))
+		g.Expect(createdMachine.Namespace).To(Equal(kcp.Namespace))
+		g.Expect(createdMachine.OwnerReferences).To(HaveLen(1))
+		g.Expect(createdMachine.OwnerReferences).To(ContainElement(*metav1.NewControllerRef(kcp, controlplanev1.GroupVersion.WithKind("KubeadmControlPlane"))))
+		g.Expect(createdMachine.Spec).To(Equal(expectedMachineSpec))
+
+		// Verify that the machineTemplate.ObjectMeta has been propagated to the Machine.
+		for k, v := range kcpMachineTemplateObjectMeta.Labels {
+			g.Expect(createdMachine.Labels[k]).To(Equal(v))
+		}
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/helpers_test.go:591]: Let's assert the entire map, same for annotations below

Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/helpers_test.go:591]: (Please also check other places that we touch in this PR)

File: controlplane/kubeadm/internal/controllers/helpers_test.go [hunk 13]
```diff
@@ -539,53 +555,132 @@ func TestKubeadmControlPlaneReconciler_generateMachine(t *testing.T) {
...
+		machineUID := types.UID("abc-123-existing-machine")
+		clusterConfigurationString := "cluster-configuration-information"
+		failureDomain := pointer.String("fd-1")
+		machineVersion := pointer.String("v1.25.3")
+		exitingMachine := &clusterv1.Machine{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: machineName,
+				UID:  machineUID,
+				Annotations: map[string]string{
+					controlplanev1.KubeadmClusterConfigurationAnnotation: clusterConfigurationString,
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/helpers_test.go:623]: let's also add the remediation annotation  (would have found the bug I found there)

File: controlplane/kubeadm/internal/controllers/scale_test.go [hunk 14]
```diff
@@ -23,7 +23,10 @@ import (
 	"time"
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/scale_test.go:23]: Q: Is it possible / is the plan to move the changes in scale_test.go, upgrade_test.go and the unrelated ones in controller_test.go in a separate PR?

This PR is already big and I would like to avoid having those changes here if they are not strictly necessary

(I'm struggling to find the time to review all of this in a timely manner)

File: controlplane/kubeadm/internal/controllers/controller_test.go [hunk 15]
```diff
@@ -1227,88 +1257,471 @@ kubernetesVersion: metav1.16.1`,
...
+	kcp.Spec.MachineTemplate.NodeVolumeDetachTimeout = duration10s
+
+	// Use the updated KCP.
+	controlPlane.KCP = kcp
+	g.Expect(reconciler.syncMachines(ctx, controlPlane)).To(Succeed())
+
+	// Verify in-place mutable fields are updated on the Machine.
+	updatedInplaceMutatingMachine = inPlaceMutatingMachine.DeepCopy()
+	g.Eventually(func(g Gomega) {
+		g.Expect(env.GetAPIReader().Get(ctx, client.ObjectKeyFromObject(updatedInplaceMutatingMachine), updatedInplaceMutatingMachine))
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/controller_test.go:1657]: Do we need the eventually here and below? Seems everything should work without it

File: controlplane/kubeadm/internal/controllers/controller_test.go [hunk 16]
```diff
@@ -1227,88 +1257,471 @@ kubernetesVersion: metav1.16.1`,
...
+			Not(BeNil()),
+			HaveValue(Equal(*kcp.Spec.MachineTemplate.NodeDrainTimeout)),
+		))
+		g.Expect(updatedInplaceMutatingMachine.Spec.NodeDeletionTimeout).Should(And(
+			Not(BeNil()),
+			HaveValue(Equal(*kcp.Spec.MachineTemplate.NodeDeletionTimeout)),
+		))
+		g.Expect(updatedInplaceMutatingMachine.Spec.NodeVolumeDetachTimeout).Should(And(
+			Not(BeNil()),
+			HaveValue(Equal(*kcp.Spec.MachineTemplate.NodeDrainTimeout)),
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/controller_test.go:1673]:
```suggestion
			HaveValue(Equal(*kcp.Spec.MachineTemplate. NodeVolumeDetachTimeout)),
```

File: controlplane/kubeadm/internal/controllers/controller_test.go [hunk 17]
```diff
@@ -1227,88 +1257,471 @@ kubernetesVersion: metav1.16.1`,
...
+		g.Expect(updatedKubeadmConfig.GetLabels()).Should(Equal(expectedLabels))
+		// Verify Annotations
+		g.Expect(updatedKubeadmConfig.GetAnnotations()).Should(Equal(kcp.Spec.MachineTemplate.ObjectMeta.Annotations))
+		// Verify spec remains the same
+		g.Expect(updatedKubeadmConfig.Spec).Should(Equal(existingKubeadmConfig.Spec))
+	}, timeout).Should(Succeed())
+
+	// Wait to ensure Machine is not updated.
+	// Verify that the machine stays the same consistently.
+	g.Consistently(func(g Gomega) {
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/controller_test.go:1703]: Same here. Seems we don't need it as we call SyncMachines directly

File: controlplane/kubeadm/internal/controllers/helpers.go (hunk 10 above)
Comment (ykakarap) [reply controlplane/kubeadm/internal/controllers/helpers.go:404]: Good catch!

File: controlplane/kubeadm/internal/controllers/scale_test.go (hunk 14 above)
Comment (ykakarap) [reply controlplane/kubeadm/internal/controllers/scale_test.go:23]: The changes in these files are necessary. The tests would fails otherwise, because the underlying code is now adopted to SSA. 
TLDR; all these tests eventually end up calling `createMachine` which is now using SSA and therefore would fail if we are not using SSA in the corresponding tests. All the other changes in made to the test are needed once we move to SSA.

Note: Setting `disableInPlacePropagation` to true would not help as `createMachine` is untouched by this flag.

File: controlplane/kubeadm/internal/controllers/controller_test.go (hunk 15 above)
Comment (ykakarap) [reply controlplane/kubeadm/internal/controllers/controller_test.go:1657]: You are right, since the test uses APIReader we dont need to use an Eventually block. 👍🏼

File: controlplane/kubeadm/internal/controllers/scale_test.go (hunk 14 above)
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/scale_test.go:23]: Alright, makes sense. Thx!

File: controlplane/kubeadm/internal/controllers/controller_test.go [hunk 18]
```diff
@@ -1227,88 +1257,460 @@ kubernetesVersion: metav1.16.1`,
...
+			Kind:       "KubeadmControlPlane",
+			APIVersion: controlplanev1.GroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			UID:       types.UID("abc-123-control-plane"),
+			Name:      "existing-kcp",
+			Namespace: namespace.Name,
+		},
+		Spec: controlplanev1.KubeadmControlPlaneSpec{
+			Version:           "v1.25.3",
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/controllers/controller_test.go:1524]: nit: would it be possible to use a different version here? Just to also verify that this field is not in-place mutated.

(please also check other non-in-place-mutable fields, should be at least failureDomain and providerID)

File: controlplane/kubeadm/internal/controllers/scale_test.go [hunk 19]
```diff
@@ -32,22 +35,66 @@ import (
...
+		t.Log("Deleting the namespace")
+		g.Expect(env.Delete(ctx, ns)).To(Succeed())
+	}
 
-	fakeClient := newFakeClient(cluster.DeepCopy(), kcp.DeepCopy(), genericMachineTemplate.DeepCopy())
+	g := NewWithT(t)
+	namespace := setup(t, g)
+	defer teardown(t, g, namespace)
+
+	cluster, kcp, _ := createClusterWithControlPlane(namespace.Name)
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/scale_test.go:65]: Took another look. Can you please check how much effort it would be to just already modify the `createClusterWithControlPlane` func to the new type?

I think it could just work even in the other cases and then we can avoid a lot of new code (defining the new object everywhere + modifying the infra ref)

**Comment by sbueringer** [comment]
Just a few nits and only in tests and error messages

Nice work!!

/assign @fabriziopandini

File: controlplane/kubeadm/internal/controllers/scale_test.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (ykakarap) [reply controlplane/kubeadm/internal/controllers/scale_test.go:65]: Checked it and luckily refactoring this to use the InfraMachineTemplate from the builder was simple enough. Switched it.

File: controlplane/kubeadm/internal/controllers/upgrade_test.go [hunk 20]
```diff
@@ -115,17 +172,24 @@ func TestKubeadmControlPlaneReconciler_RolloutStrategy_ScaleUp(t *testing.T) {
...
-	g.Expect(fakeClient.List(ctx, finalMachine, client.InNamespace(cluster.Namespace))).To(Succeed())
-	g.Expect(finalMachine.Items).To(HaveLen(1))
-
-	// assert that the deleted machine is the oldest, initial machine
-	g.Expect(finalMachine.Items[0].Name).ToNot(Equal(initialMachine.Items[0].Name))
-	g.Expect(finalMachine.Items[0].CreationTimestamp.Time).To(BeTemporally(">", initialMachine.Items[0].CreationTimestamp.Time))
+	g.Eventually(func(g Gomega) {
+		g.Expect(env.List(ctx, finalMachine, client.InNamespace(cluster.Namespace))).To(Succeed())
+		g.Expect(finalMachine.Items).To(HaveLen(1))
+		// assert that the deleted machine is the oldest, initial machine
```
[the remark this answers is not in the document]
Comment (ykakarap) [reply controlplane/kubeadm/internal/controllers/upgrade_test.go:190]: The test still checks that the old machine is deleted and the new machine is kept around by checking the name.

I had to drop the creationTimestamp based additional check because now that we switched to envtest for this test the real apiserver apparently does not have enough resolution to time for this test. Because of how quickly the test runs the creationTimestamps of the new and old machines are equal in most runs. 
Note that since we are verifying that the final list of machines is of length 1 and that the only machine is not the old machine we can guarantee that the old machine is the one that got deleted. Therefore it is safe to drop the creationTimestamp check here.

File: controlplane/kubeadm/internal/controllers/controller_test.go (hunk 18 above)
Comment (ykakarap) [reply controlplane/kubeadm/internal/controllers/controller_test.go:1524]: Added some more checks to make sure that other parts of the spec are not updated.

File: controlplane/kubeadm/internal/controllers/controller.go [hunk 21]
```diff
@@ -77,6 +80,12 @@ type KubeadmControlPlaneReconciler struct {
 
 	managementCluster         internal.ManagementCluster
 	managementClusterUncached internal.ManagementCluster
+
+	// disableInPlacePropagation should only be used for tests. This is used to skip
+	// some parts of the controller that need SSA as the current test setup does not
+	// support SSA. This flag should be dropped after all affected tests are migrated
+	// to envtest.
+	disableInPlacePropagation bool
```
Comment (ykakarap) [inline controlplane/kubeadm/internal/controllers/controller.go:88]: Added a follow-up item in the umbrella issue to drop this and migrate the affected tests to envtest.

Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller.go:88]: 👍

File: controlplane/kubeadm/internal/controllers/upgrade_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/upgrade_test.go]: Sounds good. Thx for the explanation.

I thought either we should assert it again or change the comment. Changing the comment makes sense!

File: controlplane/kubeadm/internal/controllers/controller_test.go (hunk 18 above)
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller_test.go]: Thx!

File: controlplane/kubeadm/internal/controllers/scale_test.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/scale_test.go]: Nice! Less noise on this PR and less tech debt :)

**Comment by sbueringer** [comment]
Thank you very much! Great work and clean implementation.

And as in the previous PRs small improvements across the board in terms of general quality (go doc, tests, ...) 
Nice!

/lgtm

**Comment by fabriziopandini** [comment]
Great work!
/lgtm
/approve
